pkg/ccl/changefeedccl/kvevent/blocking_buffer.go

```diff
@@ -61,27 +70,32 @@ func (b *blockingBuffer) pop() (e *bufferEntry, closed bool) {
 }
 
 // Get implements kvevent.Reader interface.
-func (b *blockingBuffer) Get(ctx context.Context) (ev Event, err error) {
+func (b *blockingBuffer) Get(ctx context.Context) (ev Event, resource Resource, err error) {
 	for {
 		got, closed := b.pop()
 		if closed {
-			return Event{}, nil
+			return Event{}, NoResource, nil
 		}
 
 		if got != nil {
 			e := got.e
 			e.bufferGetTimestamp = timeutil.Now()
-			b.qp.Update(func(r quotapool.Resource) (shouldNotify bool) {
-				res := r.(*blockingBufferQuotaPool)
-				res.release(got)
-				return true
-			})
-			return e, nil
+			resource := &callbackResource{
+				cb: func() {
+					b.qp.Update(func(r quotapool.Resource) (shouldNotify bool) {
+						res := r.(*blockingBufferQuotaPool)
+						res.release(got)
+						return true
+					})
+				},
```

Comment:
I'd prefer if you just built yourself a data type like `quotapool.IntAlloc`. It'd be literally exactly like that. This callback thing doesn't feel like it's giving you all that much. If you make a data type you can also stick a `Merge` method on it to accumulate inside the sink which should also allow for way less synchronization.